op.go

```diff
@@ -15,42 +15,231 @@
 package jq
 
 import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"reflect"
 	"strings"
-
-	"github.com/savaki/jq/scanner"
 )
 
 // Op defines a single transformation to be applied to a []byte
 type Op interface {
-	Apply([]byte) ([]byte, error)
+	Apply(reflect.Value) (reflect.Value, error)
 }
 
 // OpFunc provides a convenient func type wrapper on Op
-type OpFunc func([]byte) ([]byte, error)
+type OpFunc func(reflect.Value) (reflect.Value, error)
 
 // Apply executes the transformation defined by OpFunc
-func (fn OpFunc) Apply(in []byte) ([]byte, error) {
+func (fn OpFunc) Apply(in reflect.Value) (reflect.Value, error) {
 	return fn(in)
 }
 
+func callChain(v reflect.Value, chainFun ...Op) (reflect.Value, error) {
+	for _, f := range chainFun {
+		var err error
+		v, err = f.Apply(v)
+		if err != nil {
+			return v, err
+		}
+	}
+	return v, nil
+}
+
 // Dot extract the specific key from the map provided; to extract a nested value, use the Dot Op in conjunction with the
 // Chain Op
-func Dot(key string) OpFunc {
+func Dot(key string, chainFun ...Op) OpFunc {
 	key = strings.TrimSpace(key)
 	if key == "" {
-		return func(in []byte) ([]byte, error) { return in, nil }
+		return func(in reflect.Value) (reflect.Value, error) { return callChain(in, chainFun...) }
+	}
+
+	return func(in reflect.Value) (reflect.Value, error) {
+		for in.Kind() == reflect.Interface || in.Kind() == reflect.Ptr {
+			in = in.Elem()
+		}
+		switch in.Kind() {
+		case reflect.Map:
+			var err error
+			mapVal := in.MapIndex(reflect.ValueOf(key))
+			newMapVal := reflect.New(in.MapIndex(reflect.ValueOf(key)).Type())
```

Comment:
Why not using `mapVal` here ? (`reflect.New(mapVal)...`)

Reply:
No reason, it is even better for map for interface as we can obtain the concrete type for the key.